sys/shell/commands/sc_netif.c

```diff
@@ -137,37 +137,52 @@ void _netif_addr(int argc, char **argv)
         puts("error: invalid interface given");
         return;
     }
-    /* if address was given, parse and set it */
-    if (argc >= 3) {
-        addr_len = _parse_addr(argv[2], addr);
-        if (addr_len == 0) {
-            puts("error: unable to parse address");
-        }
-        else {
-            res = ng_netapi_set(dev, NETCONF_OPT_ADDRESS, 0,
-                                     addr, addr_len);
-            if (res > 0) {
-                printf("success: address of interface %i set to ", dev);
-                _print_addr(addr, addr_len);
-                puts("");
-            }
-            else {
-                puts("error: could not set address\n");
-            }
-        }
+    addr_len = (size_t)atoi(argv[2]);
+    if (addr_len > MAX_ADDR_LEN) {
+        puts("error: given address size is too large\n");
+        return;
+    }
+    res = ng_netapi_get(dev, NETCONF_OPT_ADDRESS, 0, addr, addr_len);
+    if (res == addr_len) {
+        printf("address of interface %i is ", dev);
+        _print_addr(addr, addr_len);
+        puts("");
     }
-    /* otherwise read it */
     else {
-        res = ng_netapi_get(dev, NETCONF_OPT_ADDRESS, 0, addr, MAX_ADDR_LEN);
-        if (res > 0) {
-            addr_len = (size_t)res;
-            printf("address of interface %i is ", dev);
-            _print_addr(addr, addr_len);
-            puts("");
-        }
-        else {
-            puts("error: unable to read address");
-        }
+        puts("error: unable to read address");
+    }
+}
+
+void _netif_addr_set(int argc, char **argv)
+{
+    kernel_pid_t dev;
+    uint8_t addr[MAX_ADDR_LEN];
+    size_t addr_len;
+    int res;
+
+    if (argc < 3) {
+        printf("usage: %s IF ADDR\n", argv[0]);
+        return;
+    }
+    /* parse interface */
+    dev = (kernel_pid_t)atoi(argv[1]);
+    if (thread_get(dev) == NULL) {
+        puts("error: invalid interface given");
+        return;
+    }
+    addr_len = _parse_addr(argv[2], addr);
+    if (addr_len == 0) {
+        puts("error: unable to parse address");
+        return;
+    }
+    res = ng_netapi_set(dev, NETCONF_OPT_ADDRESS, 0, addr, addr_len);
+    if (res == addr_len) {
```

Comment:
The meaning of the `len` parameter is quite straight forward in my opinion. When calling `set`, you specify the length of the given address (as you state in your example above). For interfaces, that support addresses of different length, this parameter can be used to detect which address should be set (short vs. long address). So it is possible to set both addresses with the same function.

When calling `get`, the `max_len` parameter is used in the same way. If I would just give the `MAX_ADDR_LEN`, the driver would have no means of finding out, which address is actually requested. By setting `max_len` to say 2 or 8, the driver can read from this context, which address the user wants to read.

So for `set`, the `len` parameter defines the exact length of the data to be written to the device. For `get`, the `max_len` parameter defines for one the maximum number of bytes that can be written to the data buffer, but gives also a context for the amount of data that is expected.

Clearer now?